staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/controller.go

```diff
@@ -231,90 +264,115 @@ func (c *celAdmissionController) Validate(
+						paramKind.String()), definition, binding)
 					continue
 				}
 
-				if len(paramsNamespace) == 0 {
-					param, err = paramInfo.controller.Informer().Get(paramsName)
+				if len(paramRef.Namespace) == 0 {
+					param, err = paramInfo.controller.Informer().Get(paramRef.Name)
 				} else {
-					param, err = paramInfo.controller.Informer().Namespaced(paramsNamespace).Get(paramsName)
+					param, err = paramInfo.controller.Informer().Namespaced(paramRef.Namespace).Get(paramRef.Name)
 				}
 
 				if err != nil {
 					// Apply failure policy
 					addConfigError(err, definition, binding)
 
-					if k8serrors.IsNotFound(err) {
-						// Param doesnt exist yet?
-						// Maybe just have to wait a bit.
+					if k8serrors.IsInvalid(err) {
+						// Param mis-configured
+						// require to set paramRef.namespace for namespaced resource and unset paramRef.namespace for cluster scoped resource
 						continue
 					}
 
 					// There was a bad internal error
 					utilruntime.HandleError(err)
 					continue
 				}
+			} else {
+				paramRef := binding.Spec.ParamRef
+				if paramRef != nil {
+					// return error if ValidatingAdmissionPolicyBinding is mis-configured
+					return fmt.Errorf("ValidatingAdmissionPolicyBinding '%s' is not allowed to set paramRef to '%s' since policyName refers to a ValidatingAdmissionPolicy '%s' which has no paramKind",
+						binding.Name, paramRef.String(), definition.Name)
+				}
 			}
 
-			if bindingInfo.evaluator == nil {
+			validator := bindingInfo.validator.Load()
+			if validator == nil {
 				// Compile policy definition using binding
-				bindingInfo.evaluator, err = definition.Compile(c.objectConverter, c.restMapper)
-				if err != nil {
-					// compilation error. Apply failure policy
-					wrappedError := fmt.Errorf("failed to compile CEL expression: %w", err)
-					addConfigError(wrappedError, definition, binding)
-					continue
-				}
-				c.bindingInfos[namespacedBindingName] = bindingInfo
+				newValidator := c.validatorCompiler.Compile(definition)
+				validator = &newValidator
+
+				bindingInfo.validator.Store(validator)
+			}
+
+			decisions, err := (*validator).Validate(a, o, param, matchKind)
+			if err != nil {
+				// runtime error. Apply failure policy
+				wrappedError := fmt.Errorf("failed to evaluate CEL expression: %w", err)
+				addConfigError(wrappedError, definition, binding)
+				continue
 			}
 
-			decisions := bindingInfo.evaluator(a, param)
 			for _, decision := range decisions {
-				switch decision.Kind {
-				case Admit:
-					// Do nothing
-				case Deny:
-					allDecisions = append(allDecisions, PolicyDecisionWithMetadata{
-						Definition:     definition,
-						Binding:        binding,
-						PolicyDecision: decision,
+				switch decision.kind {
+				case admit:
+					if len(decision.message) != 0 {
+						klog.Info("ignored ValidatingAdmissionPolicy %w failure for binding %w, due to FailurePolicy=Ignore. Error: %w", definition.Name, binding.Name, decision.message)
```

Comment:
follow-up: same comment about turning this into metrics and not logging

Reply:
Tracked as follow-up